### PR TITLE
Feature/event interceptor

### PR DESCRIPTION
**Pull request by mtrupkin**
Common event logging infrastructure.

### CODE REVIEW

File: src/main/java/org/opentestsystem/delivery/logging/EventLoggerBase.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  /**
+   * Concrete classes are required to supply the application name they are logging from.
+   * @return name of the application that is logging
+   */
+  protected abstract String getApp();
+
+  public void info(final EventInfo eventInfo) {
+    info(getApp(), eventInfo.event(), eventInfo.checkpoint().or(""), null, eventInfo.data());
```
Comment (trapnine) [inline src/main/java/org/opentestsystem/delivery/logging/EventLoggerBase.java]: Will this null throw away any potential message in the eventInfo? (ed: fixed in subsequent push adding withMessage() - please review)

File: src/main/java/org/opentestsystem/delivery/logging/EventParser.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+public class EventParser {
+
+  /**
+   * The default pre request parser.
+   *
+   * @param request
+   * @return
+   */
+  public Optional<EventInfo> parsePreHandle(final HttpServletRequest request, final Object handler, final EventLogger logger) {
+    return Optional.of(EventInfo.create(request.getPathInfo(), getEventDataFields(request)));
```
Comment (trapnine) [inline src/main/java/org/opentestsystem/delivery/logging/EventParser.java:27]: The parsePostHandle below sets the EXIT checkpoint - should this one in parsePreHandle set the ENTER checkpoint?

File: src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  public boolean preHandle(HttpServletRequest request,
+                           HttpServletResponse response, Object handler) throws Exception {
+    final EventParser parser = factory.get(request);
+    final Optional<EventInfo> eventInfo = parser.parsePreHandle(request, handler, logger);
+    if (eventInfo.isPresent()) {
+      logger.info(eventInfo.get().withCheckpoint(ENTER.name()));
+      // TODO: make sure these are needed - logger should be injected, and contains a field stash.
+      request.setAttribute(EVENT_INFO, eventInfo.get());
+      request.setAttribute(LOGGER_NAME, logger);
+    }
```
Comment (trapnine) [inline src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java]: @mtrupkin , Re: the TODO note I made above - do we need the logger and eventInfo put into the request as attributes? I haven't used this method of access yet in student - did you need it in proctor? This might be leftover from our old plan. Either way let's address then remove the TODO note.

**Review by trapnine — CHANGES_REQUESTED** [review]
A few notes, minor stuff, looks good.

File: pom.xml [hunk 4]
```diff
@@ -287,6 +287,13 @@
             <version>15.0</version>
         </dependency>
```
Comment (trapnine) [inline pom.xml:288]: Will we release this as 3.0.1 now, and have proctor point at 3.0.1? (and would that require all dependent projects using this to be bumped to point at 3.0.1 also, and would that bump be done in a different commit/review?)

File: src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+  @Autowired
+  public LoggingInterceptor(final EventParserFactory factory, final EventLogger logger) {
+    this.factory = factory;
+    this.logger = logger;
+  }
+
+  @Override
+  public boolean preHandle(HttpServletRequest request,
+                           HttpServletResponse response, Object handler) throws Exception {
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java:35]: Will standard SLF4J logs also be bound to the request/event context?  Maybe via [MDC](https://www.slf4j.org/api/org/slf4j/MDC.html)?

File: src/main/java/org/opentestsystem/delivery/logging/EventParser.java [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+   * @param request
+   * @return
+   */
+  public Optional<EventInfo> parsePostHandle(final HttpServletRequest request, final HttpServletResponse response, final Object
+    handler, EventLogger logger) {
+    return Optional.of(EventInfo.create(request.getPathInfo(), EventLogger.Checkpoint.EXIT.name(), getEventDataFields
+      (request)));
+  }
+
+  public static Map<EventLogger.EventData, Object> getEventDataFields(final HttpServletRequest request) {
```
Comment (wtritch) [inline src/main/java/org/opentestsystem/delivery/logging/EventParser.java]: nit: I'd love javadocs on all the public non-\@Override methods =)

File: src/main/java/org/opentestsystem/delivery/logging/EventParser.java (hunk 2 above)
Comment (mtrupkin) [reply src/main/java/org/opentestsystem/delivery/logging/EventParser.java:27]: The check points are being explicitly set in the interceptor (and exception handlers) after the calls to parsePreHandle and parsePostHandle.  I'll remove setting the EXIT checkpoint in the EventParser.

File: pom.xml (hunk 4 above)
Comment (mtrupkin) [reply pom.xml:288]: After the pull request for the SS_SharedCode repository is merged into develop branch, we will release/version it.  We will then make sure proctor and student are dependent on that released version in their respective develop branch.

File: src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java (hunk 3 above)
Comment (wtritch) [inline src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java]: I wonder if we want parser.parsePreHandle to instead perform the logging and request attribute setting rather than just return EventInfo? So this interceptor just becomes an EventParser lookup and handoff... What do you think?

File: src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java (hunk 5 above)
Comment (mtrupkin) [reply src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java:35]: There is an existing [Servlet Filter](https://github.com/SmarterApp/SS_SharedMultiJar/blob/82ba1490747dda2f64ae2a1343f5089431d506fa/shared-web/src/main/java/AIR/Common/Web/Session/HttpRequestLoggerInitializerFilter.java) that leverages MDC.  Normal slf4j logging in the web apps have data added from that filter.

File: src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java (hunk 3 above)
Comment (mtrupkin) [reply src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java]: I do like the idea and the point was discussed.  I'm open to changing the parser to have the side-effect of writing the log.

Comment (trapnine) [reply src/main/java/org/opentestsystem/delivery/logging/LoggingInterceptor.java]: I like that, it makes the logging behavior easy to override in the custom parsers.

Recommend adding methods to the base EventParser that get called by the interceptor (perhaps just preHandle and postHandle) that can be overridden by the custom EventParsers. These new methods can call the existing parsePreHandle/parsePostHandle, which don't have to change, then do the logging. Implementers can choose to override all 4, or none of these methods.

**Review by trapnine — APPROVED** [review]
